### PR TITLE
Fix large sarif handling

### PR DESCRIPTION
**Pull request by marcnjaramillo**
Replace this with a description of the changes your pull request makes.

## Checklist

- [x] [CHANGELOG.md](https://github.com/github/vscode-codeql/blob/main/extensions/ql-vscode/CHANGELOG.md) has been updated to incorporate all user visible changes made by this pull request.
- [x] Issues have been created for any UI or other user-facing changes made by this pull request.
- [x] `@github/docs-content-codeql` has been cc'd in all issues for UI or other user-facing changes made by this pull request.

### CODE REVIEW

File: extensions/ql-vscode/package.json [hunk 1]
```diff
@@ -969,6 +969,8 @@
     "react": "^16.8.6",
     "react-dom": "^16.8.6",
     "semver": "~7.3.2",
+    "stream-chain": "^2.2.4",
+    "stream-json": "^1.7.3",
```
Comment (aeisenberg) [inline extensions/ql-vscode/package.json:973]: nit: For consistency:

```suggestion
    "stream-chain": "~2.2.4",
    "stream-json": "~1.7.3",
```

File: extensions/ql-vscode/package.json [hunk 2]
```diff
@@ -1003,6 +1005,8 @@
     "@types/semver": "~7.2.0",
     "@types/sinon": "~7.5.2",
     "@types/sinon-chai": "~3.2.3",
+    "@types/stream-chain": "^2.0.1",
+    "@types/stream-json": "^1.7.1",
```
Comment (aeisenberg) [inline extensions/ql-vscode/package.json:1009]:
```suggestion
    "@types/stream-chain": "~2.0.1",
    "@types/stream-json": "~1.7.1",
```

File: extensions/ql-vscode/src/cli.ts [hunk 3]
```diff
@@ -682,19 +687,56 @@ export class CodeQLCliServer implements Disposable {
...
+      const asm = Assembler.connectTo(pipeline);
+
+      // Returns a constructed Log object with the results or an empty array if no results were found.
+      // If the parser fails for any reason, it will reject the promise.
+      return await new Promise((resolve, reject) => {
+        pipeline.on('error', (error) => {
+          reject(error);
+        });
+
+          asm.on('done', (asm) => {
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/cli.ts:710]: nit: Looks like a little formatting issue here.

File: extensions/ql-vscode/src/cli.ts [hunk 4]
```diff
@@ -682,19 +687,56 @@ export class CodeQLCliServer implements Disposable {
...
+            const dummyTool : sarif.Tool = {driver: {name: ''}};
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/cli.ts:711]: Can you extract this into a top-level constant so it is only created once?

File: extensions/ql-vscode/src/cli.ts [hunk 5]
```diff
@@ -682,19 +687,56 @@ export class CodeQLCliServer implements Disposable {
...
+            } else {
+              const log : sarif.Log = {
+                version: '2.1.0', 
+                runs: [
+                  { 
+                    tool: dummyTool, 
+                    results: []
+                  }
+                ]
+              };
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/cli.ts:733]: Here's a simpler solution (but the formatting is off):

```suggestion
              const log : sarif.Log = {
                version: '2.1.0', 
                runs: [
                  { 
                    tool: dummyTool, 
                    results: asm.current ?? []
                  }
                ]
              };

              resolve(log);
 ``

**Review by aeisenberg — COMMENTED** [review]
This is looking great. A few comments to clean things up.  Since this is new behaviour, I'd like to see some unit tests for this. We can chat tomorrow about this.

**Comment by aeisenberg** [comment]
I _think_ the test failures are because you are using an older version of npm. Could you try upgrading to version 8?

**Comment by aeisenberg** [comment]
Actually, try rebasing on this PR https://github.com/github/vscode-codeql/pull/1003

The problem is that for some reason, fsevents is _not_ marked as optional.

**Review by TrapzAbit — APPROVED** [review]

File: extensions/ql-vscode/src/vscode-tests/no-workspace/cli.test.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+  });
+
+  it('should return an empty array if there are no results', async () => {
+    const result = await CodeQLCliServer.parseSarif(__dirname + '/data/sarif/emptyResultsSarif.sarif');
+    expect(result.runs[0].results?.length).to.eq(0);
+  });
+
+  it('should throw an error if the file fails to parse', async () => {
+    const result = await CodeQLCliServer.parseSarif(__dirname + '/data/sarif/invalidSarif.sarif');
+    await expect(result).to.rejectedWith(/Error: Parsing output of interpretation failed: /);
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/vscode-tests/no-workspace/cli.test.ts:22]: Semantics are the same, but reads a little nicer.

```suggestion
    await expect(result).to.be.rejectedWith(/Error: Parsing output of interpretation failed: /);
```

File: extensions/ql-vscode/src/vscode-tests/no-workspace/cli.test.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+import { CodeQLCliServer } from '../../cli';
+
+chai.use(chaiAsPromised);
+const expect = chai.expect;
+
+describe.only('cliServerTests', () => {
+  it('should parse a valid SARIF file', async () => {
+    const result = await CodeQLCliServer.parseSarif(__dirname + '/data/sarif/validSarif.sarif');
+    expect(result.runs.length).to.eq(1);
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/vscode-tests/no-workspace/cli.test.ts:12]: I think you want a different expectation here.

Comment (marcnjaramillo) [reply extensions/ql-vscode/src/vscode-tests/no-workspace/cli.test.ts]: I updated this. Hopefully it's a better test.

File: extensions/ql-vscode/src/cli.ts [hunk 8]
```diff
@@ -576,6 +582,46 @@ export class CodeQLCliServer implements Disposable {
     }
   }
 
+  static async parseSarif(interpretedResultsPath: string) : Promise<sarif.Log> {
```
Comment (adityasharad) [inline extensions/ql-vscode/src/cli.ts:585]: This function doesn't require the CodeQL CLI. I think a better place for it would be `sarif-utils.ts`. Then you can place your tests for it in `sarif-utils.test.ts` and make them "pure" unit tests, i.e. completely independent of the CodeQL CLI and VS Code API. This is nice because it shows that you're testing the JSON parsing behaviour in isolation, regardless of how it integrates into CodeQL commands and the rest of the extension.

File: extensions/ql-vscode/CHANGELOG.md [hunk 9]
```diff
@@ -3,6 +3,8 @@
 ## [UNRELEASED]
 
 - Fix the _CodeQL: Open Referenced File_ command for Windows systems. [#979](https://github.com/github/vscode-codeql/pull/979)
+- Fix the _CodeQL: Open Referenced File_ command for Windows systems. [#979](https://github.com/github/vscode-codeql/pull/979) 
+- Fix a bug that causes VSCode to crash when handling large SARIF files (>4GB) [#1004](https://github.com/github/vscode-codeql/pull/1004)
```
Comment (adityasharad) [inline extensions/ql-vscode/CHANGELOG.md:7]: I would rephrase this more proactively. You've rewritten it to overcome a limitation of Node.js defaults, rather than fixing a bug in our handling.
```suggestion
- Support large SARIF results files (>4GB) without crashing VS Code [#1004](https://github.com/github/vscode-codeql/pull/1004)
```

File: extensions/ql-vscode/src/vscode-tests/no-workspace/cli.test.ts [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+const expect = chai.expect;
+
+describe.only('cliServerTests', function() {
+
+  it('should parse a valid SARIF file', async () => {
+    const result = await CodeQLCliServer.parseSarif(__dirname + '/data/sarif/validSarif.sarif');
+    expect(result.version).to.exist;
+    expect(result.runs).to.exist;
+    expect(result.runs[0].tool).to.exist;
+    expect(result.runs[0].tool.driver).to.exist;
```
Comment (adityasharad) [inline extensions/ql-vscode/src/vscode-tests/no-workspace/cli.test.ts:16]: Add an assertion for the number of results?

**Review by adityasharad — COMMENTED** [review]
This looks promising! Some suggestions on organisation.

File: extensions/ql-vscode/CHANGELOG.md [hunk 11]
```diff
@@ -3,6 +3,8 @@
 ## [UNRELEASED]
 
 - Fix the _CodeQL: Open Referenced File_ command for Windows systems. [#979](https://github.com/github/vscode-codeql/pull/979)
+- Fix the _CodeQL: Open Referenced File_ command for Windows systems. [#979](https://github.com/github/vscode-codeql/pull/979)
```
Comment (aeisenberg) [inline extensions/ql-vscode/CHANGELOG.md:6]: Oops. this line is duplicated. Looks like a bad merge.

File: extensions/ql-vscode/src/cli.ts (hunk 8 above)
Comment (aeisenberg) [reply extensions/ql-vscode/src/cli.ts:585]: Yes. That makes more sense. It was originally my suggestion to make this function static.

**Review by aeisenberg — APPROVED** [review]
Nice work!
